### PR TITLE
Add makefile convention

### PR DESCRIPTION
**Pull request by skjnldsv**
Please help :)
- [ ] Appstore script
- [ ] Signature integration ?
- [x] Npm scripts
- [ ] Testing?
- [ ] Other?

### CODE REVIEW

File: developer_manual/app/makefile.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
+.. sectionauthor:: John Molakvoæ <skjnldsv@protonmail.com>
+.. codeauthor:: John Molakvoæ <skjnldsv@protonmail.com>
+..  _makefile:
+
+========
+Makefile
+========
+
+To ease our developement, we aime to have the same script naming convention.
```
Comment (MorrisJobke) [inline developer_manual/app/makefile.rst]:
```suggestion
To ease our developement, we aim to have the same script naming convention.
```

**Comment by MorrisJobke** [comment]
Wasn't the idea to use krankerl by @ChristophWurst?

**Comment by skjnldsv** [comment]
I'm fine with whatever we chose as long as it's the same for our apps :)

**Comment by MorrisJobke** [comment]
> I'm fine with whatever we chose as long as it's the same for our apps :)

Problem with that: another tool that needs to be installed - problem with Makefile: quite some maintenance effort to keep them up to date.

**Comment by skjnldsv** [comment]
So, what shall we do @MorrisJobke ? :)

File: developer_manual/app/makefile.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
+.. sectionauthor:: John Molakvoæ <skjnldsv@protonmail.com>
+.. codeauthor:: John Molakvoæ <skjnldsv@protonmail.com>
+..  _makefile:
+
+========
+Makefile
+========
+
+To ease our developement, we aim to have the same script naming convention.
```
Comment (MorrisJobke) [inline developer_manual/app/makefile.rst]:
```suggestion
To ease our development, we aim to have the same script naming convention.
```

**Comment by MorrisJobke** [comment]
> So, what shall we do @MorrisJobke ? :)

I'm fine with this for now.

**Comment by jancborchardt** [comment]
Btw, when I do make, there’s a bunch of changed files, which are always a hassle to stash. Is there a plan for that? :)

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   apps/accessibility/js/accessibility.js
	modified:   apps/accessibility/js/accessibility.js.map
	modified:   apps/twofactor_backupcodes/js/settings.js
	modified:   apps/twofactor_backupcodes/js/settings.js.map
	modified:   apps/updatenotification/js/updatenotification.js
	modified:   apps/updatenotification/js/updatenotification.js.map
	modified:   apps/workflowengine/appinfo/info.xml
	modified:   apps/workflowengine/js/workflowengine.js
	modified:   apps/workflowengine/js/workflowengine.js.map
	modified:   settings/js/vue-0.js
	modified:   settings/js/vue-0.js.map
	modified:   settings/js/vue-4.js
	modified:   settings/js/vue-4.js.map
	modified:   settings/js/vue-5.js
	modified:   settings/js/vue-5.js.map
	modified:   settings/js/vue-6.js
	modified:   settings/js/vue-6.js.map
	modified:   settings/js/vue-settings-admin-security.js
	modified:   settings/js/vue-settings-admin-security.js.map
	modified:   settings/js/vue-settings-apps-users-management.js
	modified:   settings/js/vue-settings-apps-users-management.js.map
	modified:   settings/js/vue-settings-personal-security.js
	modified:   settings/js/vue-settings-personal-security.js.map

no changes added to commit (use "git add" and/or "git commit -a")
jan@Rechenknecht:~/nextcloud$ git checkout master
error: Your local changes to the following files would be overwritten by checkout:
	apps/updatenotification/js/updatenotification.js
	apps/updatenotification/js/updatenotification.js.map
	apps/workflowengine/appinfo/info.xml
	settings/js/vue-0.js
	settings/js/vue-0.js.map
	settings/js/vue-settings-admin-security.js
	settings/js/vue-settings-admin-security.js.map
	settings/js/vue-settings-apps-users-management.js
	settings/js/vue-settings-apps-users-management.js.map
	settings/js/vue-settings-personal-security.js
	settings/js/vue-settings-personal-security.js.map
Please commit your changes or stash them before you switch branches.
Aborting
```

**Comment by skjnldsv** [comment]
@jancborchardt `make clean`, `git stash` :)

**Comment by jancborchardt** [comment]
> @jancborchardt `make clean`, `git stash` :)

As said, `which are always a hassle to stash` ;) – but yeah, if this is the only way then it should be documented in the readme.

**Comment by skjnldsv** [comment]
> As said, `which are always a hassle to stash` ;)

If you do a make clean, they're not a hassle to stash, it will never leave anything left. That's why I gave you this command :)

**Comment by ChristophWurst** [comment]
> If you do a make clean, they're not a hassle to stash, it will never leave anything left. That's why I gave you this command :)

Makefiles are so simple, right.

/me hides

**Comment by skjnldsv** [comment]
@ChristophWurst this have nothing to do with makefiles though, it's just a shortand to remove files.
The same issue would happens with something like krankerl :)

**Comment by jancborchardt** [comment]
I understand all that and of course there’s several terminal commands to fix the situation. But I’m saying that either this should not happen _or_ it should be fixed automatically. :D

**Comment by skjnldsv** [comment]
> But I’m saying that either this should not happen _or_ it should be fixed automatically. :D

Do a bug report to git then ^^

**Comment by jancborchardt** [comment]
We control the makefiles, this is an issue literally about the makefile convention, we could add an operation to fix this to the makefiles. Or is that not correct?

**Comment by juliushaertl** [comment]
> We control the makefiles, this is an issue literally about the makefile convention, we could add an operation to fix this to the makefiles. Or is that not correct?

I would say that makes sense. I'm not sure how the command should be called since make clean is already in use but maybe `make clear-build` which then just runs `git checkout -- js/`?

**Comment by skjnldsv** [comment]
I feel like it's a bit too overpowered, but sure, we could indeed do something! You're 100% right @jancborchardt :wink: 
@ChristophWurst does krankerl have something like that?


PS: before, it could be nice to review this pr and discuss further improvements later, it's been already 4 months xD

**Review by juliushaertl — APPROVED** [review]
Fine by me :+1:

**Comment by juliushaertl** [comment]
@skjnldsv I'd say we get this in after a rebase as it is just documenting how we do it and makes it easier for app developers to follow our convention :+1:

**Comment by skjnldsv** [comment]
> @skjnldsv I'd say we get this in after a rebase as it is just documenting how we do it and makes it easier for app developers to follow our convention

The more I think about this, the more I think we're not easing the devs workflow here.
Creating proprietary shortcuts with makefiles that doesn't reflect any standard will just make newcomer's knowledge be biased on Nextcloud only. They should learn the npm basics and git basics instead. Otherwise we're just making it more complicated for them. 

I don't use makefiles anymore and would actually be in favour of removing them.

**Comment by juliushaertl** [comment]
Also fine by me. 

@jancborchardt @ChristophWurst @ma12-co What do you think? Shall we just close this then?

**Comment by ChristophWurst** [comment]
> The more I think about this, the more I think we're not easing the devs workflow here.
> Creating proprietary shortcuts with makefiles that doesn't reflect any standard will just make newcomer's knowledge be biased on Nextcloud only. They should learn the npm basics and git basics instead. Otherwise we're just making it more complicated for them.

Couldn't agree more. That's why I was always against them.

**Comment by MorrisJobke** [comment]
> The more I think about this, the more I think we're not easing the devs workflow here.
> Creating proprietary shortcuts with makefiles that doesn't reflect any standard will just make newcomer's knowledge be biased on Nextcloud only. They should learn the npm basics and git basics instead. Otherwise we're just making it more complicated for them.

Makes sense 👍

**Pull request closed without merging**